"Error: no snapshots found for buffer" when renaming type in Go source file

### Check for existing issues

- [x] Completed

### Describe the bug / provide steps to reproduce it

- I've opened a Go project with zed.
- gopls server is setup and have been working fine in general.
- When I try to rename a type, I get the error "no snapshots found for buffer 30064771180 and server 1"
- The type is renamed on the file itself, but all references in other files aren't updated.

### Zed Version and System Specs

Zed: v0.170.2 (Zed)
OS: Linux Wayland ubuntu 24.10
Memory: 15.2 GiB
Architecture: x86_64
GPU: Intel(R) Iris(R) Plus Graphics (ICL GT2) || Intel open-source Mesa driver || Mesa 24.2.3-1ubuntu1

### If applicable, add screenshots or screencasts of the incorrect state / behavior

![Image](https://github.com/user-attachments/assets/72c74029-387f-4d4f-b777-7c164daa7b96)

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary>

<!-- Click below this line and paste or drag-and-drop your log-->
```
2025-01-27T12:09:45.264354688Z [INFO] ========== starting zed ==========
2025-01-27T12:09:45.292811727Z [INFO] Presentation support: "VK_KHR_wayland_surface"
2025-01-27T12:09:45.292901101Z [INFO] Presentation support: "VK_KHR_xcb_surface"
2025-01-27T12:09:45.292929411Z [INFO] Presentation support: "VK_KHR_xlib_surface"
2025-01-27T12:09:45.29296026Z [INFO] Enabling Vulkan Portability
2025-01-27T12:09:45.292986873Z [INFO] Enabling color space support
2025-01-27T12:09:45.322128652Z [INFO] Adapter: "Intel(R) Iris(R) Plus Graphics (ICL GT2)"
2025-01-27T12:09:45.322243319Z [INFO] No ray tracing extensions are supported
2025-01-27T12:09:45.324125743Z [INFO] perform;
2025-01-27T12:09:45.324212303Z [INFO] authenticate;
2025-01-27T12:09:45.324258653Z [INFO] Opening main db
2025-01-27T12:09:45.324310938Z [INFO] read_command;
2025-01-27T12:09:45.324429498Z [INFO] send_secondary_commands;
2025-01-27T12:09:45.324854319Z [INFO] new;
2025-01-27T12:09:45.324899595Z [INFO] socket reader;
2025-01-27T12:09:45.325134495Z [INFO] Opening main db
2025-01-27T12:09:45.325470802Z [INFO] keep_updated;
2025-01-27T12:09:45.327525135Z [INFO] Using git binary path: None
2025-01-27T12:09:45.364237226Z [INFO] No prompt template overrides directory found at /home/fornellas/.local/share/zed/prompt_overrides. Using built-in prompts.
2025-01-27T12:09:45.364816456Z [INFO] No prompt template overrides directory found at /home/fornellas/.local/share/zed/prompt_overrides. Using built-in prompts.
2025-01-27T12:09:45.37244792Z [INFO] extensions updated. loading 3, reloading 0, unloading 0
2025-01-27T12:09:45.374183854Z [INFO] activate is not implemented on Linux, ignoring the call
2025-01-27T12:09:45.375770976Z [INFO] perform;
2025-01-27T12:09:45.375822855Z [INFO] authenticate;
2025-01-27T12:09:45.375882172Z [INFO] read_command;
2025-01-27T12:09:45.376552549Z [INFO] socket reader;
2025-01-27T12:09:45.381765488Z [INFO] new;
2025-01-27T12:09:45.382316379Z [INFO] Opening main db
2025-01-27T12:09:45.383300815Z [INFO] keep_updated;
2025-01-27T12:09:45.38729641Z [INFO] set status on client 0: Authenticating
2025-01-27T12:09:45.387571316Z [INFO] perform;
2025-01-27T12:09:45.387603162Z [INFO] authenticate;
2025-01-27T12:09:45.387638985Z [INFO] read_command;
2025-01-27T12:09:45.38826594Z [INFO] socket reader;
2025-01-27T12:09:45.388812997Z [INFO] building git repository, `.git` path in the worktree: ".git"
2025-01-27T12:09:45.388921562Z [INFO] Using surface present mode MAILBOX
2025-01-27T12:09:45.38989251Z [INFO] Creating a descriptor pool for at most 16 sets
2025-01-27T12:09:45.390046229Z [INFO] Creating a descriptor pool for at most 16 sets
2025-01-27T12:09:45.39017694Z [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: PreMultiplied }
2025-01-27T12:09:45.393466632Z [INFO] new;
2025-01-27T12:09:45.394449842Z [INFO] keep_updated;
2025-01-27T12:09:45.401641185Z [INFO] Using GPU: GpuSpecs { is_software_emulated: false, device_name: "Intel(R) Iris(R) Plus Graphics (ICL GT2)", driver_name: "Intel open-source Mesa driver", driver_info: "Mesa 24.2.3-1ubuntu1" }
2025-01-27T12:09:45.402206326Z [INFO] set status on client 248442: Connecting
2025-01-27T12:09:45.403761507Z [INFO] Using surface present mode MAILBOX
2025-01-27T12:09:45.403805726Z [INFO] Retaining current format: Bgra8UnormSrgb
2025-01-27T12:09:45.404077182Z [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: Ignored }
2025-01-27T12:09:45.418172526Z [INFO] Using surface present mode MAILBOX
2025-01-27T12:09:45.418217482Z [INFO] Retaining current format: Bgra8UnormSrgb
2025-01-27T12:09:45.418423192Z [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: PreMultiplied }
2025-01-27T12:09:45.42845607Z [INFO] Using surface present mode MAILBOX
2025-01-27T12:09:45.428499828Z [INFO] Retaining current format: Bgra8UnormSrgb
2025-01-27T12:09:45.429778013Z [INFO] Using surface present mode MAILBOX
2025-01-27T12:09:45.429829995Z [INFO] Retaining current format: Bgra8UnormSrgb
2025-01-27T12:09:45.446841761Z [INFO] Opening main db
2025-01-27T12:09:45.510307177Z [INFO] using project environment variables from CLI. PATH="/home/fornellas/.cache/service-software-catalog-connector/GOROOT/go1.23.5.linux-amd64/bin:/home/fornellas/.cache/service-software-catalog-connector/kubeconform/linux/amd64/bin:/home/fornellas/.cache/service-software-catalog-connector/helm/v3.15.3/linux-amd64:/home/fornellas/.cache/service-software-catalog-connector/GOROOT/go1.23.5.linux-amd64/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/fornellas/.uc/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/fornellas/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin"
2025-01-27T12:09:45.535502965Z [WARN] unhandled message: Metadata(SupermavenMetadataMessage { dust_strings: None })
2025-01-27T12:09:45.61542037Z [INFO] connected to rpc endpoint https://collab.zed.dev/rpc
2025-01-27T12:09:45.713342853Z [INFO] set environment variables from shell:/bin/bash, path:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/fornellas/.cache/service-software-catalog-connector/GOROOT/go1.23.5.linux-amd64/bin:/home/fornellas/.cache/service-software-catalog-connector/kubeconform/linux/amd64/bin:/home/fornellas/.cache/service-software-catalog-connector/helm/v3.15.3/linux-amd64:/home/fornellas/.cache/service-software-catalog-connector/GOROOT/go1.23.5.linux-amd64/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/fornellas/.uc/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/fornellas/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin
2025-01-27T12:09:45.987605849Z [INFO] add_connection;
2025-01-27T12:09:45.993042551Z [INFO] set status on client 248442: Connected { peer_id: PeerId { owner_id: 670, id: 3599856 }, connection_id: ConnectionId { owner_id: 0, id: 0 } }
2025-01-27T12:09:46.078534506Z [INFO] Using surface present mode MAILBOX
2025-01-27T12:09:46.078591532Z [INFO] Retaining current format: Bgra8UnormSrgb
2025-01-27T12:09:46.137352885Z [INFO] Using surface present mode MAILBOX
2025-01-27T12:09:46.137392764Z [INFO] Retaining current format: Bgra8UnormSrgb
2025-01-27T12:09:46.30713171Z [INFO] Using surface present mode MAILBOX
2025-01-27T12:09:46.30717779Z [INFO] Retaining current format: Bgra8UnormSrgb
2025-01-27T12:09:56.947999875Z [INFO] Opening main db
2025-01-27T12:09:56.984200712Z [INFO] attempting to start language server "gopls", path: "/home/fornellas/src/service-software-catalog-connector", id: 1
2025-01-27T12:09:56.985342287Z [INFO] using project environment variables from CLI. PATH="/home/fornellas/.cache/service-software-catalog-connector/GOROOT/go1.23.5.linux-amd64/bin:/home/fornellas/.cache/service-software-catalog-connector/kubeconform/linux/amd64/bin:/home/fornellas/.cache/service-software-catalog-connector/helm/v3.15.3/linux-amd64:/home/fornellas/.cache/service-software-catalog-connector/GOROOT/go1.23.5.linux-amd64/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/fornellas/.uc/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/home/fornellas/bin:/home/linuxbrew/.linuxbrew/bin:/home/linuxbrew/.linuxbrew/sbin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/usr/games:/usr/local/games:/snap/bin"
2025-01-27T12:09:56.985742417Z [INFO] fetching latest version of language server "gopls"
2025-01-27T12:09:56.986352235Z [WARN] unhandled message: Set(SupermavenSetMessage { key: "disabled", value: Bool(false) })
2025-01-27T12:09:57.43904943Z [INFO] downloading language server "gopls"
2025-01-27T12:09:57.447013067Z [INFO] starting language server process. binary path: "/home/fornellas/.local/share/zed/languages/gopls/gopls_0.17.1_go_1.23.5", working directory: "/home/fornellas/src/service-software-catalog-connector", args: ["-mode=stdio"]
2025-01-27T12:09:57.897979875Z [WARN] unhandled capability registration: Registration { id: "workspace/didChangeConfiguration", method: "workspace/didChangeConfiguration", register_options: None }
2025-01-27T12:10:15.822032036Z [INFO] Creating a descriptor pool for at most 16 sets
2025-01-27T12:10:15.853957928Z [INFO] Creating a descriptor pool for at most 16 sets
2025-01-27T12:10:15.887389152Z [INFO] Creating a descriptor pool for at most 256 sets
2025-01-27T12:10:15.903008077Z [INFO] Creating a descriptor pool for at most 256 sets
2025-01-27T12:10:19.497642642Z [ERROR] no snapshots found for buffer 30064771180 and server 1
```
<!-- Click above this line and paste or drag-and-drop your log--></details>

--- a/crates/project/src/lsp_store.rs
+++ b/crates/project/src/lsp_store.rs
@@ -1986,13 +1986,21 @@ impl LocalLspStore {
 
         if let Some(version) = version {
             let buffer_id = buffer.read(cx).remote_id();
-            let snapshots = self
+            let snapshots = if let Some(snapshots) = self
                 .buffer_snapshots
                 .get_mut(&buffer_id)
                 .and_then(|m| m.get_mut(&server_id))
-                .ok_or_else(|| {
-                    anyhow!("no snapshots found for buffer {buffer_id} and server {server_id}")
-                })?;
+            {
+                snapshots
+            } else if version == 0 {
+                // Some language servers report version 0 even if the buffer hasn't been opened yet.
+                // We detect this case and treat it as if the version was `None`.
+                return Ok(buffer.read(cx).text_snapshot());
+            } else {
+                return Err(anyhow!(
+                    "no snapshots found for buffer {buffer_id} and server {server_id}"
+                ));
+            };
 
             let found_snapshot = snapshots
                     .binary_search_by_key(&version, |e| e.version)

